File: ios/Cove/Flows/SendFlow/SendFlowManager.swift
//
//  SendFlowManager.swift
//  Cove
//
//  Created by Praveen Perera on 4/24/25.
//

import Foundation
import os
import SwiftUI

extension WeakReconciler: SendFlowManagerReconciler where Reconciler == SendFlowManager {}

protocol SendFlowRustManaging: AnyObject, Sendable {
    func walletId() -> WalletId
    func enteringFiatAmount() -> String
    func sendAmountFiat() -> String
    func sendAmountBtc() -> String
    func totalSpentInFiat() -> String
    func totalSpentInBtc() -> String
    func totalFeeString() -> String?
    func listenForUpdates(reconciler: SendFlowManagerReconciler)
    func validateAddress(displayAlert: Bool) -> Bool
    func validateAmount(displayAlert: Bool) -> Bool
    func getCustomFeeOption(
        feeRate: FeeRate,
        feeSpeed: FeeSpeed
    ) async throws -> FeeRateOptionWithTotalFee
    func waitForInit() async -> Bool
    func amountExceedsBalance() -> Bool
    func sanitizeBtcEnteringAmount(oldValue: String, newValue: String) -> String?
    func sanitizeFiatEnteringAmount(oldValue: String, newValue: String) -> String?
    func utxos() -> [Utxo]?
    func maxSendMinusFees() -> Amount?
    func maxSendMinusFeesAndSmallUtxo() -> Amount?
    func dispatch(action: SendFlowManagerAction)
}

extension RustSendFlowManager: SendFlowRustManaging {}

private enum SendFlowManagerAccessError: LocalizedError {
    case closed

    var errorDescription: String? {
        "Send flow manager is closed"
    }
}

@Observable final class SendFlowManager: ReconcilingManager, SendFlowManagerReconciler {
    typealias Message = SendFlowManagerReconcileMessage
    typealias Action = SendFlowManagerAction

    private struct RustState {
        var rust: SendFlowRustManaging? = nil
        var debouncedTask: Task<Void, Never>? = nil
        var delayedAlertWorkItem: DispatchWorkItem? = nil
        var isClosed = false
    }

    private let logger = Log(id: "SendFlowManager")
    @ObservationIgnored
    private let rustState = OSAllocatedUnfairLock(initialState: RustState())
    @ObservationIgnored
    private let rustBridge = DispatchQueue(
        label: "cove.SendFlowManager.rustbridge", qos: .userInitiated
    )

    @ObservationIgnored
    let id: WalletId

    private var rust: SendFlowRustManaging? {
        rustState.withLock { $0.rust }
    }

    var enteringBtcAmount: String = ""
    var enteringFiatAmount: String = ""
    private var _enteringAddress: String = ""

    var address: Address? = nil
    var amount: Amount? = nil
    var fiatAmount: Double? = nil

    var presenter: SendFlowPresenter
    var feeSelection: FeeSelection? = nil
    var selectedFeeRate: FeeRateOptionWithTotalFee? {
        feeSelection?.selected
    }

    var feeRateOptions: FeeRateOptionsWithTotalFee? {
        feeSelection?.options
    }

    var maxSelected: Amount? = nil

    // presenting
    var sendAmountFiat: String = ""
    var sendAmountBtc: String = ""

    var totalSpentInFiat: String = ""
    var totalSpentInBtc: String = ""
    var totalFeeString: String? = nil

    var enteringAddress: Binding<String> {
        Binding<String>(
            get: { self._enteringAddress },
            set: { newValue in
                guard self.canApplyReconcileMessages else { return }

                self._enteringAddress = newValue
                self.dispatch(action: .notifyEnteringAddressChanged(newValue))
            }
        )
    }

    func validate(displayAlert: Bool = false) -> Bool {
        validateAmount(displayAlert: displayAlert)
            && validateAddress(displayAlert: displayAlert)
    }

    func validateAddress(displayAlert: Bool = false) -> Bool {
        withRustOr(false) { $0.validateAddress(displayAlert: displayAlert) }
    }

    func validateAmount(displayAlert: Bool = false) -> Bool {
        withRustOr(false) { $0.validateAmount(displayAlert: displayAlert) }
    }

    init(_ rust: SendFlowRustManaging, presenter: SendFlowPresenter) {
        self.presenter = presenter

        self.id = rust.walletId()
        self.enteringFiatAmount = rust.enteringFiatAmount()
        self.sendAmountFiat = rust.sendAmountFiat()
        self.sendAmountBtc = rust.sendAmountBtc()
        self.totalSpentInFiat = rust.totalSpentInFiat()
        self.totalSpentInBtc = rust.totalSpentInBtc()
        self.totalFeeString = rust.totalFeeString()

        rustState.withLock { $0.rust = rust }
        rust.listenForUpdates(reconciler: WeakReconciler(self))
    }

    deinit {
        close()
    }

    func close() {
        let state = takeRustAndQueuedWorkForClose()
        guard state.rust != nil else { return }

        state.debouncedTask?.cancel()
        state.delayedAlertWorkItem?.cancel()
        logger.debug("Closed SendFlowManager for wallet \(id)")
    }

    private func takeRustAndQueuedWorkForClose() -> (
        rust: SendFlowRustManaging?,
        debouncedTask: Task<Void, Never>?,
        delayedAlertWorkItem: DispatchWorkItem?
    ) {
        rustState.withLock { state in
            guard !state.isClosed else { return (nil, nil, nil) }

            state.isClosed = true
            let rust = state.rust
            let debouncedTask = state.debouncedTask
            let delayedAlertWorkItem = state.delayedAlertWorkItem
            state.rust = nil
            state.debouncedTask = nil
            state.delayedAlertWorkItem = nil
            return (rust, debouncedTask, delayedAlertWorkItem)
        }
    }

    private func requireRust() throws -> SendFlowRustManaging {
        guard let rust else { throw SendFlowManagerAccessError.closed }
        return rust
    }

    private func withRustOr<T>(
        _ defaultValue: T,
        _ body: (SendFlowRustManaging) -> T
    ) -> T {
        guard let rust else { return defaultValue }
        return body(rust)
    }

    private func installDebouncedTask(_ task: Task<Void, Never>) {
        let previousTask = rustState.withLock { state -> Task<Void, Never>? in
            guard !state.isClosed else {
                task.cancel()
                return nil
            }

            let previousTask = state.debouncedTask
            state.debouncedTask = task
            return previousTask
        }

        previousTask?.cancel()
    }

    private func installDelayedAlertWorkItem(_ workItem: DispatchWorkItem?) {
        let previousWorkItem = rustState.withLock { state -> DispatchWorkItem? in
            guard !state.isClosed else {
                workItem?.cancel()
                return nil
            }

            let previousWorkItem = state.delayedAlertWorkItem
            state.delayedAlertWorkItem = workItem
            return previousWorkItem
        }

        previousWorkItem?.cancel()
    }

    public func setAddress(_ address: Address) {
        guard canApplyReconcileMessages else { return }

        self._enteringAddress = address.unformatted()
        self.address = address
        self.dispatch(action: .notifyAddressChanged(address))
    }

    public func setAmount(_ amount: Amount) {
        guard canApplyReconcileMessages else { return }

        self.amount = amount
        self.dispatch(action: .notifyAmountChanged(amount))
    }

    public func refreshPresenters() {
        guard let rust else { return }

        self.totalSpentInFiat = rust.totalSpentInFiat()
        self.totalSpentInBtc = rust.totalSpentInBtc()
        self.totalFeeString = rust.totalFeeString()
        self.sendAmountBtc = rust.sendAmountBtc()
        self.sendAmountFiat = rust.sendAmountFiat()
    }

    public func reconcileAfterLabelsChanged() {
        dispatch(action: .refreshWalletBalance)
    }

    public func getNewCustomFeeRateWithTotal(
        feeRate: FeeRate, feeSpeed: FeeSpeed
    ) async throws -> FeeRateOptionWithTotalFee {
        let rust = try requireRust()
        return try await rust.getCustomFeeOption(
            feeRate: feeRate, feeSpeed: feeSpeed
        )
    }

    func waitForInit() async -> Bool {
        guard let rust else { return false }
        return await rust.waitForInit()
    }

    func amountExceedsBalance() -> Bool {
        withRustOr(false) { $0.amountExceedsBalance() }
    }

    func sanitizeBtcEnteringAmount(oldValue: String, newValue: String) -> String? {
        withRustOr(nil) {
            $0.sanitizeBtcEnteringAmount(oldValue: oldValue, newValue: newValue)
        }
    }

    func sanitizeFiatEnteringAmount(oldValue: String, newValue: String) -> String? {
        withRustOr(nil) {
            $0.sanitizeFiatEnteringAmount(oldValue: oldValue, newValue: newValue)
        }
    }

    func utxos() -> [Utxo]? {
        withRustOr(nil) { $0.utxos() }
    }

    func maxSendMinusFees() -> Amount? {
        withRustOr(nil) { $0.maxSendMinusFees() }
    }

    func maxSendMinusFeesAndSmallUtxo() -> Amount? {
        withRustOr(nil) { $0.maxSendMinusFeesAndSmallUtxo() }
    }

    var canApplyReconcileMessages: Bool {
        rust != nil
    }

    func apply(_ message: Message) {
        guard canApplyReconcileMessages else { return }

        switch message {
        case let .updateAmountFiat(fiat):
            self.fiatAmount = fiat

        case let .updateAmountSats(sats):
            self.refreshPresenters()
            self.amount = Amount.fromSat(sats: sats)

        case let .updateFeeSelection(selection):
            self.refreshPresenters()
            self.feeSelection = selection

        case let .updateAddress(address):
            self.address = address

        case let .updateEnteringBtcAmount(amount):
            self.enteringBtcAmount = amount

        case let .updateEnteringAddress(address):
            self._enteringAddress = address

        case let .updateEnteringFiatAmount(amount):
            self.enteringFiatAmount = amount

        case let .updateFocusField(field):
            self.presenter.focusField = field

        case let .setAlert(alertState):
            applySetAlert(alertState)

        case .clearAlert:
            installDelayedAlertWorkItem(nil)
            self.presenter.clearAlert()

        case let .setMaxSelected(maxSelected):
            self.maxSelected = maxSelected

        case .unsetMaxSelected:
            self.maxSelected = nil

        case .refreshPresenters:
            self.refreshPresenters()
        }
    }

    private func applySetAlert(_ alertState: SendFlowAlertState) {
        Log.warn("setAlert: \(alertState)")
        let hadSheet = presenter.sheetState != .none
        let hadAlert = presenter.alertState != .none
        let isDismissingAlert = presenter.isDisappearing

        if hadSheet || hadAlert || isDismissingAlert {
            presenter.clearAlert()
            presenter.sheetState = .none

            let workItem = DispatchWorkItem { [weak self] in
                guard let self, self.canApplyReconcileMessages else { return }

                self.presenter.alertState = .init(alertState)
            }
            installDelayedAlertWorkItem(workItem)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.6, execute: workItem)
        } else {
            installDelayedAlertWorkItem(nil)
            presenter.alertState = .init(alertState)
        }
    }

    func logReconcile(message: Message) {
        logger.debug("reconcile: \(message)")
    }

    func logReconcileMany(messages: [Message]) {
        logger.debug("reconcile_messages: \(messages)")
    }

    public func dispatch(action: Action) {
        dispatch(action)
    }

    public func dispatch(_ action: Action) {
        rustBridge.async { [weak self] in
            guard let self, let rust = self.rust else { return }

            self.logger.debug("dispatch: \(action)")
            rust.dispatch(action: action)
        }
    }

    public func debouncedDispatch(
        _ action: Action, for debounceDelay: Duration? = .milliseconds(66)
    ) {
        guard let debounceDelay else { return self.dispatch(action) }

        let task = Task { [weak self] in
            guard let self else { return }
            try? await Task.sleep(for: debounceDelay)
            guard !Task.isCancelled else { return }
            self.dispatch(action)
        }
        installDebouncedTask(task)
    }
}

File: android/app/src/main/java/org/bitcoinppl/cove/flows/SendFlow/SendFlowManager.kt
package org.bitcoinppl.cove.flows.SendFlow

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.bitcoinppl.cove.RustHandleGuard
import org.bitcoinppl.cove.TaggedItem
import org.bitcoinppl.cove_core.*
import org.bitcoinppl.cove_core.types.*
import java.io.Closeable
import java.util.concurrent.atomic.AtomicBoolean

/**
 * send flow manager - manages send transaction flow state
 * ported from iOS SendFlowManager.swift
 */
@Stable
class SendFlowManager internal constructor(
    private val rust: RustSendFlowManager,
    var presenter: SendFlowPresenter,
) : SendFlowManagerReconciler,
    Closeable {
    private val tag = "SendFlowManager"

    // Scope for UI-bound work; reconcile and UI updates run on Main
    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val isClosed = AtomicBoolean(false)
    private val rustGuard =
        RustHandleGuard(
            ownerName = "SendFlowManager",
            handleName = "RustSendFlowManager",
            isClosed = isClosed,
        ) {
            logWarn(it)
        }

    val id: WalletId = rust.walletId()

    // user input state
    var enteringBtcAmount by mutableStateOf("")
        private set

    var enteringFiatAmount by mutableStateOf("")
        private set

    private var _enteringAddress by mutableStateOf("")

    // validated state
    var address by mutableStateOf<Address?>(null)

    var amount by mutableStateOf<Amount?>(null)

    var fiatAmount by mutableStateOf<Double?>(null)
        private set

    // fee state
    var feeSelection by mutableStateOf<FeeSelection?>(null)
        private set

    val selectedFeeRate: FeeRateOptionWithTotalFee?
        get() = feeSelection?.selected

    val feeRateOptions: FeeRateOptionsWithTotalFee?
        get() = feeSelection?.options

    var maxSelected by mutableStateOf<Amount?>(null)
        private set

    // presenting strings
    var sendAmountFiat by mutableStateOf("")
        private set

    var sendAmountBtc by mutableStateOf("")
        private set

    var totalSpentInFiat by mutableStateOf("")
        private set

    var totalSpentInBtc by mutableStateOf("")
        private set

    var totalFeeString by mutableStateOf<String?>(null)
        private set

    // debounce task
    private var debouncedTask: Job? = null

    init {
        logDebug("Initializing SendFlowManager for $id")
        enteringFiatAmount = rust.enteringFiatAmount()
        sendAmountFiat = rust.sendAmountFiat()
        sendAmountBtc = rust.sendAmountBtc()
        totalSpentInFiat = rust.totalSpentInFiat()
        totalSpentInBtc = rust.totalSpentInBtc()
        totalFeeString = rust.totalFeeString()

        rust.listenForUpdates(this)
    }

    private fun logDebug(message: String) {
        android.util.Log.d(tag, message)
    }

    private fun logWarn(message: String) {
        android.util.Log.w(tag, message)
    }

    private suspend fun <T> withRustSuspend(
        block: suspend RustSendFlowManager.() -> T,
    ): T = rustGuard.withHandleSuspend(rust, block)

    private fun <T> withRustOr(
        defaultValue: T,
        block: RustSendFlowManager.() -> T,
    ): T = rustGuard.withHandleOr(rust, defaultValue, block)

    private suspend fun <T> withRustOrSuspend(
        defaultValue: T,
        block: suspend RustSendFlowManager.() -> T,
    ): T = rustGuard.withHandleOrSuspend(rust, defaultValue, block)

    /**
     * get/set entering address with dispatch
     */
    var enteringAddress: String
        get() = _enteringAddress
        set(value) {
            if (isClosed.get()) return
            _enteringAddress = value
            dispatch(SendFlowManagerAction.NotifyEnteringAddressChanged(value))
        }

    /**
     * update entering BTC amount with debounced dispatch
     * only dispatches if value actually changed (matches iOS pattern)
     */
    fun updateEnteringBtcAmount(value: String) {
        if (isClosed.get()) return
        if (enteringBtcAmount != value) {
            enteringBtcAmount = value
            debouncedDispatch(SendFlowManagerAction.NotifyEnteringBtcAmountChanged(value))
        }
    }

    /**
     * update entering fiat amount with debounced dispatch
     * only dispatches if value actually changed (matches iOS pattern)
     */
    fun updateEnteringFiatAmount(value: String) {
        if (isClosed.get()) return
        if (enteringFiatAmount != value) {
            enteringFiatAmount = value
            debouncedDispatch(SendFlowManagerAction.NotifyEnteringFiatAmountChanged(value))
        }
    }

    /**
     * validate entire send flow
     */
    fun validate(displayAlert: Boolean = false): Boolean {
        if (isClosed.get()) return false
        return validateAmount(displayAlert) && validateAddress(displayAlert)
    }

    suspend fun waitForInit(): Boolean =
        withRustOrSuspend(false) {
            waitForInit()
        }

    fun amountExceedsBalance(): Boolean =
        withRustOr(false) {
            amountExceedsBalance()
        }

    fun currentAmount(): Amount? =
        withRustOr(null) {
            amount()
        }

    fun maxSendMinusFees(): Amount? =
        withRustOr(null) {
            maxSendMinusFees()
        }

    fun maxSendMinusFeesAndSmallUtxo(): Amount? =
        withRustOr(null) {
            maxSendMinusFeesAndSmallUtxo()
        }

    fun sanitizeBtcEnteringAmount(
        oldValue: String,
        newValue: String,
    ): String? =
        withRustOr(null) {
            sanitizeBtcEnteringAmount(oldValue, newValue)
        }

    fun sanitizeFiatEnteringAmount(
        oldValue: String,
        newValue: String,
    ): String? =
        withRustOr(null) {
            sanitizeFiatEnteringAmount(oldValue, newValue)
        }

    fun validateAddress(displayAlert: Boolean = false): Boolean =
        withRustOr(false) {
            validateAddress(displayAlert)
        }

    fun validateAmount(displayAlert: Boolean = false): Boolean =
        withRustOr(false) {
            validateAmount(displayAlert)
        }

    fun updateAddress(address: Address) {
        if (isClosed.get()) return
        _enteringAddress = address.unformatted()
        this.address = address
        dispatch(SendFlowManagerAction.NotifyAddressChanged(address))
    }

    fun updateAmount(amount: Amount) {
        if (isClosed.get()) return
        this.amount = amount
        dispatch(SendFlowManagerAction.NotifyAmountChanged(amount))
    }

    fun refreshPresenters() {
        totalSpentInFiat =
            withRustOr(totalSpentInFiat) {
                totalSpentInFiat()
            }
        totalSpentInBtc =
            withRustOr(totalSpentInBtc) {
                totalSpentInBtc()
            }
        totalFeeString =
            withRustOr(totalFeeString) {
                totalFeeString()
            }
        sendAmountBtc =
            withRustOr(sendAmountBtc) {
                sendAmountBtc()
            }
        sendAmountFiat =
            withRustOr(sendAmountFiat) {
                sendAmountFiat()
            }
    }

    fun reconcileAfterLabelImport() {
        dispatch(SendFlowManagerAction.RefreshWalletBalance)
    }

    suspend fun getNewCustomFeeRateWithTotal(
        feeRate: FeeRate,
        feeSpeed: FeeSpeed,
    ): FeeRateOptionWithTotalFee =
        withRustSuspend {
            getCustomFeeOption(feeRate, feeSpeed)
        }

    private fun apply(message: SendFlowManagerReconcileMessage) {
        when (message) {
            is SendFlowManagerReconcileMessage.UpdateAmountFiat -> {
                fiatAmount = message.v1
            }

            is SendFlowManagerReconcileMessage.UpdateAmountSats -> {
                refreshPresenters()
                amount = Amount.fromSat(message.v1)
            }

            is SendFlowManagerReconcileMessage.UpdateFeeSelection -> {
                refreshPresenters()
                feeSelection = message.v1
            }

            is SendFlowManagerReconcileMessage.UpdateAddress -> {
                address = message.v1
            }

            is SendFlowManagerReconcileMessage.UpdateEnteringBtcAmount -> {
                enteringBtcAmount = message.v1
            }

            is SendFlowManagerReconcileMessage.UpdateEnteringAddress -> {
                _enteringAddress = message.v1
            }

            is SendFlowManagerReconcileMessage.UpdateEnteringFiatAmount -> {
                enteringFiatAmount = message.v1
            }

            is SendFlowManagerReconcileMessage.UpdateFocusField -> {
                presenter.focusField = message.v1
            }

            is SendFlowManagerReconcileMessage.SetAlert -> {
                logWarn("setAlert: ${message.v1}")

                // capture previous state before modifying
                val hadSheet = presenter.sheetState != null
                val hadAlert = presenter.alertState != null

                presenter.alertState = TaggedItem(message.v1)

                // handle alert/sheet conflict - delay only if there was a previous conflict
                if (hadSheet || hadAlert) {
                    presenter.alertState = null
                    presenter.sheetState = null
                    mainScope.launch {
                        delay(ALERT_PRESENTATION_DELAY_MS)
                        presenter.alertState = TaggedItem(message.v1)
                    }
                }
            }

            is SendFlowManagerReconcileMessage.ClearAlert -> {
                presenter.alertState = null
            }

            is SendFlowManagerReconcileMessage.SetMaxSelected -> {
                maxSelected = message.v1
            }

            is SendFlowManagerReconcileMessage.UnsetMaxSelected -> {
                maxSelected = null
            }

            is SendFlowManagerReconcileMessage.RefreshPresenters -> {
                refreshPresenters()
            }
        }
    }

    override fun reconcile(message: SendFlowManagerReconcileMessage) {
        logDebug("reconcile: $message")
        mainScope.launch { apply(message) }
    }

    override fun reconcileMany(messages: List<SendFlowManagerReconcileMessage>) {
        logDebug("reconcile_messages: ${messages.size} messages")
        mainScope.launch { messages.forEach { apply(it) } }
    }

    fun dispatch(action: SendFlowManagerAction) {
        if (isClosed.get()) return
        logDebug("dispatch: $action")
        mainScope.launch {
            withRustOr(Unit) {
                dispatch(action)
            }
        }
    }

    /**
     * dispatch with debouncing for high-frequency updates
     */
    fun debouncedDispatch(
        action: SendFlowManagerAction,
        debounceDelayMs: Long = DEFAULT_DEBOUNCE_MS,
    ) {
        debouncedTask?.cancel()
        debouncedTask = null

        if (debounceDelayMs <= 0) {
            dispatch(action)
            return
        }

        debouncedTask =
            mainScope.launch {
                delay(debounceDelayMs)
                if (!isActive) return@launch
                dispatch(action)
            }
    }

    override fun close() {
        rustGuard.closeOnce {
            logDebug("Closing SendFlowManager for $id")
            debouncedTask?.cancel()
            debouncedTask = null
            mainScope.cancel()
            rust.close()
        }
    }

    companion object {

        /**
         * delay before showing alert when another modal (sheet/alert) was visible
         *
         * allows previous modal dismiss animation to complete before presenting a new alert
         * material3 bottom sheet dismiss animation ≈ 500ms → extra buffer prevents flicker
         */
        private const val ALERT_PRESENTATION_DELAY_MS = 600L

        /**
         * default debounce for amount input
         * ~60fps target = 16.67ms per frame, 66ms = ~4 frames
         * balances responsiveness vs rust bridge overhead
         */
        private const val DEFAULT_DEBOUNCE_MS = 66L
    }
}
